src/experimental/theoretic_information.cpp

```diff
@@ -0,0 +1,644 @@
+    float p_occ = meas_outcm[0][1];
+    float p_un = meas_outcm[0][0];
+
+    // Root
+    map_out.insert(map_pair(std::make_tuple(0, 0, 0), 1.0f));
+
+    // First measurement
+    map_out.insert(map_pair(std::make_tuple(1, 0, 0), p_free));
+    map_out.insert(map_pair(std::make_tuple(0, 1, 0), p_occ));
+    map_out.insert(map_pair(std::make_tuple(0, 0, 1), p_un));
+
+    for (int i = r; r < k; ++r)
+    {
+        std::vector<map_tuple> tup_vct;
+        std::vector<float> acc_prob;
+
+        for (it_out = map_out.begin(); it_out != map_out.end(); ++it_out)
+        {
+            // Index
+            int fr_idx = std::get<0>(it_out->first);
+            int oc_idx = std::get<1>(it_out->first);
+            int un_idx = std::get<2>(it_out->first);
+
+            // Measurement outcome probability
+            float free_prop = meas_outcm[r][0];
+            float occ_prop = meas_outcm[r][1];
+            float un_prop = meas_outcm[r][2];
+
+            if (fr_idx + oc_idx + un_idx == r)
+            {
+                // Searching for the current combination in this level
+                std::vector<map_tuple>::iterator it_comb;
+                it_comb = std::find(tup_vct.begin(), tup_vct.end(), std::make_tuple(fr_idx + 1, oc_idx, un_idx));
+
+                // Free
+                if (it_comb != tup_vct.end())
+                {
+                    acc_prob[it_comb - tup_vct.begin()] += it_out->second * free_prop;
+                }
+                else
+                {
+                    tup_vct.push_back(std::make_tuple(fr_idx + 1, oc_idx, un_idx));
+                    acc_prob.push_back(it_out->second * free_prop);
+                }
+
+                it_comb = std::find(tup_vct.begin(), tup_vct.end(), std::make_tuple(fr_idx, oc_idx + 1, un_idx));
+
+                // Occupied
+                if (it_comb != tup_vct.end())
+                {
+                    acc_prob[it_comb - tup_vct.begin()] += it_out->second * occ_prop;
+                }
+                else
+                {
+                    tup_vct.push_back(std::make_tuple(fr_idx, oc_idx + 1, un_idx));
+                    acc_prob.push_back(it_out->second * occ_prop);
+                }
+
+                it_comb = std::find(tup_vct.begin(), tup_vct.end(), std::make_tuple(fr_idx, oc_idx, un_idx + 1));
+
+                // Unobserved
+                if (it_comb != tup_vct.end())
+                {
+                    acc_prob[it_comb - tup_vct.begin()] += it_out->second * un_prop;
+                }
+                else
+                {
+                    tup_vct.push_back(std::make_tuple(fr_idx, oc_idx, un_idx + 1));
+                    acc_prob.push_back(it_out->second * un_prop);
+                }
+            }
+        }
+        // Inserting the elements into the map
+        for (int k = 0; k < tup_vct.size(); ++k)
+        {
+            map_out.insert(map_pair(tup_vct[k], acc_prob[k]));
+        }
+    }
```

Comment:
@hidmic at the beginning it was my approach; however, I noted some particular cases.

When adding the elements in the map directly, I was missing some propagations which means some probabilities were not being calculated in the right way (I believe this is due to the ordering things are being added).

I would say that this approach could be improved to reduce the iterator arithmetic, but also I think this is reliable in the way we are calculating and propagating the probabilities. I will try to find a way to avoid using some iterators